src/app/AttributePathParams.h

```diff
@@ -99,10 +99,25 @@ struct AttributePathParams
         return HasWildcardAttributeId();
     }
 
+    // check if input concrete attribute path is subset of current wildcard attribute
+    bool IncludesSingleAttributesInCluster(const ConcreteAttributePath & aOther) const
+    {
+        return IncludesAllAttributesInCluster(aOther);
+    }
+
+    // check if input attribute path is same as current one.
+    bool IsSameAttribute(const AttributePathParams & aOther) const
+    {
+        return (mEndpointId == aOther.mEndpointId) && (mClusterId == aOther.mClusterId) && (mAttributeId == aOther.mAttributeId);
+    }
+
+    bool IsDuplicate() { return mIsDuplicate; }
+
     ClusterId mClusterId     = kInvalidClusterId;   // uint32
     AttributeId mAttributeId = kInvalidAttributeId; // uint32
     EndpointId mEndpointId   = kInvalidEndpointId;  // uint16
     ListIndex mListIndex     = kInvalidListIndex;   // uint16
+    bool mIsDuplicate        = false;               // uint8
```

Comment:
What is it 'duplicated' with? Why would we ever want to add state like this to a params object like AttributePathParams? The concept of duplication is transient, and contextual to a specific use-case for a narrow window of time. Baking it pseudo-permanently like this into this object is just not a great idea, especially since this object is used in so many places in so many other contexts, none of which benefit from this.

We really should be thinking a lot deeper through some of these design decisions.